patterns/strategy.md

```diff
@@ -0,0 +1,171 @@
+# Strategy (aka Policy)
+
+## Description
+
+The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern)
+is a technique that enables separation of concerns in object-oriented software design.
+It also allows to decouple software modules through [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle).
+The basic idea behind the Strategy pattern is that, given an algorithm solving a particular problem,
+we define only the skeleton of the algorithm at an abstract level and
+we delegate the specific algorithm’s implementation (or some of its parts) to another class or method.
+
+In this way, a client using the algorithm may choose a specific implementation, while the general algorithm workflow remains the same.
+In other words, the abstract specification of the class does not depend on the specific implementation of the derived class,
+but specific implementation must adhere to the abstract specification.
+This is why we call it "Dependency Inversion".
+
+## Motivation
+
+Imagine we are working on a project that generates reports every month.
+We need the reports to be generated in different formats (strategies), e.g.,
+in `JSON` or `Plain Text` formats.
+But things vary over time and we don't know what kind of requirement we may get in the future.
+For example, we may need to generate our report in a completly new format,
+or just modify one of the existing formats.
+
+## Example
+
+In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`,
+while `Text` and `Json` are our strategy structs.
+These strategies have to implement the `Formatter` trait.
+
+```rust
+use std::collections::HashMap;
+type Data = HashMap<String, u32>;
+
+trait Formatter {
+    fn run(&self, data: &Data) -> String;
+}
+
+struct Report;
+
+impl Report {
+    fn generate<T: Formatter>(g: T) -> String {
+        // backend operations...
+        let mut data = HashMap::new();
+        data.insert("one".to_string(), 1);
+        data.insert("two".to_string(), 2);
+        // generate report
+        g.run(&data)
+    }
+}
+
+struct Text;
+impl Formatter for Text {
+    fn run(&self, data: &Data) -> String {
+        data.iter()
+            .map(|(key, val)| format!("{} {}\n", key, val))
+            .collect()
+    }
+}
+
+struct Json;
+impl Formatter for Json {
+    fn run(&self, data: &Data) -> String {
+        let mut s = String::from("[");
+        let mut entry;
+
+        for (key, val) in data {
+            if s.len() > 1 {
+                s.push(',');
+            }
+            entry = format!("{{\"{}\":\"{}\"}}", key, val);
+            s.push_str(&entry);
+        }
+        s.push(']');
+        return s;
```

Comment:
```suggestion
        let mut s = String::from("[");
        let mut iter = data.into_iter();
                            
        if let Some((key, val)) = data.next() {
            let entry = format!(r#"{"{}":"{}"}"#, key, val);
            s.push_str(&entry);
            while let Some((key, val)) = data.next() {
                s.push(',');   
                let entry = format!(r#"{"{}":"{}"}"#, key, val);
                s.push_str(&entry);
            }
        }
        s.push(']');
        s
```
This should be more efficient since we unrolled the loop to remove the branch, ideally we want to make it branchless. Like the above example, it may be a bit longer but it is more efficient.

There is also another method to this, we could collect into `Vec<String>` and `slices::join` them, although it requires double allocation which could make it shorter but less efficient maybe.

Reply:
I had in mind but didn't know how to do it in Rust. Good idiom: create iter, first next, and then iter til the end.

Reply:
not `data.next()`, but `iter.next()`